[FIRRTL] Add a bitcast operation to FIRRTL dialect

prithayan opened the pull request:
Add a new `bitcast` operation to `FIRRTL` dialect, which will be directly lowered to [hw::bitcast](https://github.com/prithayan/circt/blob/firrtlBitCastOp/docs/RationaleComb.md#bitcasts) operation.
The `bitcast` operation can be used to directly cast a `vector` or `bundle` type to a flattened `uint`, without blowing out the fields.

This operation will be used to simplify lowering of aggregate memory data type in `LowerTypes`. 
This PR also fixes the `LowerToHW` for lowering `FVectorType` to `ArrayType`.

lattner commented on docs/RationaleFIRRTL.md:
```diff
@@ -297,6 +297,11 @@ For a historical discussion of this issue and its development see:
+### `firrtl.bitcast`
+The bitcast operation represents a bitwise reinerpretation (cast) of a value. 
+This always synthesizes away in hardware, and is lowered to 
+[`hw.bitcast` ](https://github.com/prithayan/circt/blob/firrtlBitCastOp/docs/RationaleComb.md#bitcasts)
```
Please drop `https://github.com/prithayan/circt/blob/firrtlBitCastOp/docs/` from the link

prithayan replied:
Updated the doc.

lattner commented on include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:
```diff
@@ -610,4 +610,17 @@ def HWStructCastOp : FIRRTLOp<"hwStructCast", [NoSideEffect]> {
+def BitCastOp: FIRRTLOp<"bitcast", [NoSideEffect]> {
+  let summary = [{
+    Reinterpret one value to another value of the same size and
+    potentially different type. This op is lowered to hw::BitCastOp.
```
Instead of "This op is lowered to hw::BitCastOp.", how about "This op follows the same endianness policy as `hw::BitCastOp`."

lattner commented on include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:
```diff
@@ -610,4 +610,17 @@ def HWStructCastOp : FIRRTLOp<"hwStructCast", [NoSideEffect]> {
+  }];
+
+  let arguments = (ins AnyType:$input);
+  let results = (outs AnyType:$result);
```
This should use FIRRTLType instead of AnyType for both things.

prithayan replied:
Thanks, this was a typo.

lattner commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -2411,6 +2418,15 @@ LogicalResult FIRRTLLowering::visitExpr(HWStructCastOp op) {
+  auto operand = getLoweredValue(op.getOperand());
+  if (!operand)
+    return failure();
+  auto resultType = lowerType(op.getType());
```
This should check to make sure resultType isn't null.

lattner commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -2468,6 +2468,49 @@ static LogicalResult verifyHWStructCastOp(HWStructCastOp cast) {
+static LogicalResult verifyBitCastOp(BitCastOp cast) {
+  StringRef errorMsg;
+  // Infer the bitwidth of the type, if it can be computed. Else return false.
+  std::function<bool(FIRRTLType, size_t &)> getWidth =
```
Please move this to a helper function in FIRRTLTypes.h/cpp

lattner replied:
Also, please don't name it "getFoo" since it returns a bool.  "isFoo" would make more sense.

prithayan replied:
Added the function to return the width. 
A non zero width is valid, any other unknown or zero width is treated as error.

lattner commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -2468,6 +2468,49 @@ static LogicalResult verifyHWStructCastOp(HWStructCastOp cast) {
+          return true;
+        })
+        .Default([&](auto t) {
+          errorMsg = "bitcastOp only supports types with valid bit width:";
```
Why doesn't it handle clock and other things?  Should we use this instead of the existing "as clock" casts?

prithayan replied:
Right, we already have `AsSIntPrimOp`, `AsUIntPrimOp`, `AsClockPrimOp`, `AsAsyncResetPrimOp` to handle the other cases.

prithayan commented:
Thanks @lattner for the review. Addressed your comments. 
We could replace `AsUIntPrimOp` etc with this bitcast instead, I will try it out and see if it makes sense.

lattner commented:
> We could replace AsUIntPrimOp etc with this bitcast instead, I will try it out and see if it makes sense.

I think that is worth investigating, but it should definitely be a separate patch.  Thanks!

lattner commented on docs/RationaleFIRRTL.md:
```diff
@@ -297,6 +297,16 @@ For a historical discussion of this issue and its development see:
+The bitcast operation represents a bitwise reinerpretation (cast) of a value. 
+It can be used to cast a vector or bundle type to an int type or vice-versa.
+The bitwidth of input and result type must be known and non-zero. It cannot
+be used for clock types or reset types 
```
> It cannot be used for clock types or reset types 

Why not?  That seems like an arbitrary limitation, why not allow it?

prithayan replied:
Added clock and reset types.

lattner commented on include/circt/Dialect/FIRRTL/FIRRTLTypes.h:
```diff
@@ -384,6 +384,9 @@ class FVectorType : public FIRRTLType::TypeBase<FVectorType, FIRRTLType,
 };
 
+// Get the non-zero known bit width for this type, return 0 if unknown.
+size_t getBitWidth(FIRRTLType type);
```
FIRRTL supports zero bit integers, so please use ~0ULL for the sentinel.

lattner commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -2468,6 +2468,21 @@ static LogicalResult verifyHWStructCastOp(HWStructCastOp cast) {
+static LogicalResult verifyBitCastOp(BitCastOp cast) {
+
+  auto inTypeBits = getBitWidth(cast.getOperand().getType().cast<FIRRTLType>());
+  auto resTypeBits = getBitWidth(cast.getType().cast<FIRRTLType>());
```
The .cast on line 2474 shouldn't be needed anymore.

lattner commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -2468,6 +2468,21 @@ static LogicalResult verifyHWStructCastOp(HWStructCastOp cast) {
+  auto inTypeBits = getBitWidth(cast.getOperand().getType().cast<FIRRTLType>());
+  auto resTypeBits = getBitWidth(cast.getType().cast<FIRRTLType>());
+  StringRef errorMsg =
+      "the input and result must have known non-zero bit widths";
```
I'd recommend duplicating the cast.emitError to simplify the logic.  No need for an errorMsg variable.

lattner commented on lib/Dialect/FIRRTL/FIRRTLTypes.cpp:
```diff
@@ -1009,3 +1009,32 @@ void FIRRTLDialect::registerTypes() {
            CMemoryType, CMemoryPortType>();
 }
+
+size_t firrtl::getBitWidth(FIRRTLType type) {
```
Plz add doc comment.

prithayan commented:
Thanks @lattner, Added the support for clock and reset types and zero bitwidth too.

richardxia commented on include/circt/Dialect/FIRRTL/FIRRTLTypes.h:
```diff
@@ -384,6 +384,9 @@ class FVectorType : public FIRRTLType::TypeBase<FVectorType, FIRRTLType,
   std::pair<unsigned, bool> rootChildFieldID(unsigned fieldID, unsigned index);
 };
 
+// Get the bit width for this type, return None  if unknown.
```
```suggestion
// Get the bit width for this type, return None if unknown.
```

Extra space here.

richardxia commented on docs/RationaleFIRRTL.md:
```diff
@@ -297,6 +297,14 @@ For a historical discussion of this issue and its development see:
 - [`llvm/circt#992`](https://github.com/llvm/circt/pull/992)
 
+### `firrtl.bitcast`
+The bitcast operation represents a bitwise reinerpretation (cast) of a value. 
+It can be used to cast a vector or bundle type to an int type or vice-versa.
+The bitwidth of input and result type must be known.
+For an aggregate type bitwidth of every field must be known.
+This always synthesizes away in hardware, and follows the same endainness 
+policy as `hw.bitcast`.
```
```suggestion
The bitcast operation represents a bitwise reinterpretation (cast) of a value. 
It can be used to cast a vector or bundle type to an int type or vice-versa.
The bit width of input and result types must be known.
For an aggregate type, the bit width of every field must be known.
This always synthesizes away in hardware, and follows the same endianness 
policy as `hw.bitcast`.
```

Just a few spelling and grammar changes.

richardxia commented on lib/Dialect/FIRRTL/FIRRTLTypes.cpp:
```diff
@@ -1009,3 +1009,32 @@ void FIRRTLDialect::registerTypes() {
            CMemoryType, CMemoryPortType>();
 }
+
+// Get the bit width for this type, return None  if unknown.
```
```suggestion
// Get the bit width for this type, return None if unknown.
```

richardxia reviewed (commented):
I read through some of the prose and left some spelling and grammar change suggestions. I don't feel familiar enough with the code to offer a proper review of it, but it looks good to me at a high level.

darthscsi commented on docs/RationaleFIRRTL.md:
```diff
@@ -297,6 +297,14 @@ For a historical discussion of this issue and its development see:
 
+### `firrtl.bitcast`
+The bitcast operation represents a bitwise reinterpretation (cast) of a value. 
+It can be used to cast a vector or bundle type to an int type or vice-versa.
```
what about vector to/from bundle?  does it have to go through an int?

darthscsi commented on include/circt/Dialect/FIRRTL/FIRRTLTypes.h:
```diff
@@ -384,6 +384,9 @@ class FVectorType : public FIRRTLType::TypeBase<FVectorType, FIRRTLType,
 };
 
+// Get the bit width for this type, return None if unknown.
+llvm::Optional<int32_t> getBitWidth(FIRRTLType type);
```
Isn't this just getBitWidthOrSentinel?

darthscsi replied:
Ah, that sentinels out on aggregates.  Fair question, should it?  Seems redundant to have 2 almost identical things.

prithayan replied:
`getBitWidthOrSentinel` returns `-2` on an aggregate type, whereas this `getBitWidth` can recursively traverse the fields and return the total bitwidth of the leaf elements.

darthscsi commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -72,6 +72,12 @@ static Type lowerType(Type type) {
     }
     return hw::StructType::get(type.getContext(), hwfields);
   }
+  if (FVectorType vec = firType.dyn_cast<FVectorType>()) {
```
You can land this addition directly if you want.  Or as part of this, doesn't matter.

darthscsi commented on test/Dialect/FIRRTL/errors.mlir:
```diff
@@ -555,7 +555,47 @@ firrtl.circuit "MemoryPortsWithDifferentTypes" {
+firrtl.circuit "BitCast3" {
+  firrtl.module @BitCast3() {
+    %a = firrtl.wire : !firrtl.clock
+    %b = firrtl.bitcast %a : (!firrtl.clock) -> (!firrtl.uint<1>)
```
What causes this to error?  it looks like it should work (getWidth(clock) -> 1), and it seems reasonable for it to work.

darthscsi reviewed (approved):
LGTM % minor comments.  The getWidth question is the biggest open.  Does anything actually depend on getBitWidthOrSentinel returning -2 on an aggregate?

prithayan commented:
Thanks @lattner @richardxia @darthscsi for the review. Merging this PR now.
